imatrix : use GGUF to store importance matrices

compilade opened the pull request:
Follow-up from <https://github.com/ikawrakow/ik_llama.cpp/discussions/15#discussioncomment-10465793>.

Using GGUF as the format for `imatrix` files will be useful for further experiments (e.g. with [L²QER](https://github.com/ggerganov/llama.cpp/discussions/8831)) and compatibility with existing or future GGUF tooling (e.g. GGUF previews on HuggingFace, graphical GGUF viewer(s) <https://github.com/ggerganov/llama.cpp/issues/6715>, some kind of `gguf-diff`, etc.).

There are multiple problems with `imatrix` which this is addressing:

- Ad-hoc format which isn't really readable by other projects (and which has no way to backward-compatibly be extended except by adding more stuff at the end)
- Non-deterministic tensor order depending on `unordered_map` iteration order (makes `sha256sum` useless to compare `imatrix` files made on the same dataset)
- Broken behavior at small `-ub` (intermediate saves happen waaay too often)
- Can't use bigger batch size than chunk size

# Summary of changes

- Use GGUF to store `imatrix` data.
  - `general.type` is `imatrix`
  - no `general.architecture`
    - can't really know the architecture from old `imatrix` files.
  - store `*.in_sum2` and `*.counts` for each tensors with imatrix data.
    - `*.in_sum2` are the per-channel sums of squared activations
      - Stored in `F32`, like before.
    - `*.counts` are the number of activations (also the number of tokens), useful to calculate the mean squared activations (which is used by `llama-quantize`)
      - Why not simply store the mean? To allow merging `imatrix` files together with `--in-file`.
      - It's stored in `F32` even though it's integer values, because when calculating the mean it would be converted to `F32` anyway to perform the division.
- ~~Add `convert_legacy_imatrix_to_gguf.py` to convert old `imatrix.dat` files to `imatrix.gguf`~~
  - Conversion is either not necessary (since `llama-quantize` can still read the old format (with a warning)) or can be converted with `llama-imatrix` directly (when the output file has the `.gguf` suffix).
- Like `llama-perplexity` since #5946, allow computing multiple chunks per batch with `llama-imatrix`
  - This should be useful for huge models like Llama-405B when they don't fit completely in RAM.
- ~~Use fused-multiply-add (with `std::fma`) when accumulating the sums of activations~~
  - (decided against using it for now, for easier comparisons with `llama-imatrix` from `master`)
  - Shouldn't hurt to somewhat reduce rounding errors
    - (obviously `f64` would be even better, but I'm not use it's worth it yet. For the curious, using `double` for the intermediate accumulations can be tried by changing only one line in `IMatrixStats`: `vector<float> values` to `vector<double> values`.)
- Sort the tensor names before serializing
  - This makes the tensor order deterministic, because otherwise it depended on the iteration order of `unordered_map`.
    - Determinism between runs means `sha256sum` can be meaningfully used to compare `imatrix` files generated in very similar conditions.

# TODO

- [ ] Compare old `llama-quantize` with old `imatrix.dat` with new `llama-quantize` using converted `imatrix.gguf`
  - Seemed to work, but might need to re-test. The resulting quantized model(s) should have the same `sha256sum`.
- [x] Test new `llama-imatrix` at different batch sizes
  - Same checksums with `-ub 64 -b 512` and `-ub 512 -b 2048` for a chunk size of 512 (`-c 512`) 
- [x] Perplexity test(s) with i-quants with old `llama-imatrix` vs new `llama-imatrix`
- [x] Test with MoE models (perplexity with i-quants should be in the same ballpark as before)
- [ ] Test `--in-file` with `llama-imatrix`
  - [x] single `.imatrix` or `.gguf` imatrix (for round-trip conversions)
  - [ ] multiple (for merging)
- [ ] (maybe) Implement cleaner `general.architecture` exclusion.
  - Currently, this uses a subclass to make `self.add_architecture()` a no-op, but maybe `general.architecture` should simply be excluded when `self.arch == ""`. Not sure how to prevent using the other `self.add_*` (in `GGUFWriter`) which expect `self.arch` to be something.
  - Or maybe the architecture *should* be included?
    - What about conversions from older `imatrix.dat` files?

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Medium

compilade committed:
imatrix : allow processing multiple chunks per batch

* perplexity : simplify filling the batch

compilade committed:
imatrix : fix segfault when using a single chunk per batch

compilade committed:
imatrix : use GGUF to store imatrix data

compilade committed:
imatrix : fix conversion problems

compilade committed:
Merge branch 'master' into compilade/imatrix-batched-chunks

compilade committed:
imatrix : use FMA and sort tensor names

compilade committed:
py : add requirements for legacy imatrix convert script

compilade committed:
perplexity : revert changes

compilade committed:
py : include imatrix converter requirements in toplevel requirements

compilade committed:
imatrix : avoid using designated initializers in C++

compilade committed:
imatrix : remove unused n_entries

ngxson commented on examples/imatrix/imatrix.cpp:
```diff
@@ -252,102 +268,136 @@ void IMatrixCollector::save_imatrix(int ncall) const {
-                tmp[i] = (stat.values[i] / static_cast<float>(stat.counts[i])) * static_cast<float>(stat.ncall);
+            struct ggml_tensor * sums = ggml_new_tensor_2d(ctx, GGML_TYPE_F32, nval / nmat, nmat);
+            struct ggml_tensor * counts = ggml_new_tensor_2d(ctx, GGML_TYPE_F32, 1, nmat);
+            ggml_set_name(sums, (name + ".sums").c_str());
+            ggml_set_name(counts, (name + ".counts").c_str());
```
There is also `ggml_format_name` that you can use:

```suggestion
            ggml_format_name(sums, "%s.sums", name.c_str());
            ggml_format_name(counts, %s.counts", name.c_str());
```

ngxson commented on examples/imatrix/imatrix.cpp:
```diff
@@ -252,102 +268,136 @@ void IMatrixCollector::save_imatrix(int ncall) const {
+            counts_name.size() < counts_suffix.size() ||
+            !std::equal(sums_name.begin(), sums_name.end() - sums_suffix.size(), counts_name.begin()) ||
+            !std::equal(sums_suffix.rbegin(), sums_suffix.rend(), sums_name.rbegin()) ||
+            !std::equal(counts_suffix.rbegin(), counts_suffix.rend(), counts_name.rbegin())) {
```
We can probably abstract this to `str_endswith`?

```cpp
    auto str_endswith = [](const std::string & str, const std::string & suffix) {
        return str.size() >= suffix.size() && str.compare(str.size()-suffix.size(), suffix.size(), suffix) == 0;
    };
```

compilade committed:
imatrix : allow loading mis-ordered tensors

Sums and counts tensors no longer need to be consecutive.

* imatrix : more sanity checks when loading multiple imatrix files

* imatrix : use ggml_format_name instead of std::string concatenation

Co-authored-by: Xuan Son Nguyen <son@huggingface.co>

compilade committed:
quantize : use unused imatrix chunk_size with LLAMA_TRACE

compilade commented on examples/imatrix/imatrix.cpp:
```diff
@@ -252,102 +268,136 @@ void IMatrixCollector::save_imatrix(int ncall) const {
+    const std::string sums_suffix{".sums"};
+    const std::string counts_suffix{".counts"};
+
+    // TODO: allow loading from mis-ordered imatrix files
```
Good idea. I've implemented this in <https://github.com/ggerganov/llama.cpp/pull/9400/commits/8c13e16bb0f6b654ff4774e54fbc3b125ae495a6> by using a single `std::map<std::string, std::pair<struct ggml_tensor *, struct ggml_tensor *>>` and using the common prefix as the `name`, because the sums and counts tensors are iterated simultaneously in `quantize.cpp`, and this allows using a similar implementation in both `load_imatrix`.

compilade commented:
I'm setting this to "draft", because of concerns by @ikawrakow in <https://github.com/ikawrakow/ik_llama.cpp/discussions/15#discussioncomment-10615399> and <https://github.com/ikawrakow/ik_llama.cpp/discussions/15#discussioncomment-10626253> (mostly related to the fact that GGUF is harder to parse than `imatrix.dat` files).

More details near the end of <https://github.com/ikawrakow/ik_llama.cpp/discussions/15#discussioncomment-10632253>.

I'll need some days to think about how to go further with this.

compilade committed:
common : use GGUF for imatrix output by default

compilade committed:
Merge branch 'master' into compilade/imatrix-batched-chunks

compilade committed:
Merge branch 'master' into compilade/imatrix-batched-chunks

ggerganov commented:
@compilade This is a good change and I think it would be useful to bring it to a completion.

In the future, we can extend `libllama` with an interface for saving/loading `imatrix` data. This way the implementation for reading and writing the `imatrix` data would be localized in `libllama` and can be kept in-sync more easily. This can be combined with the refactoring of `llama_model_quantize_params` to [not pass C++ objects](https://github.com/ggml-org/llama.cpp/pull/12511#discussion_r2032541334).

compilade committed:
Merge branch 'master' into compilade/imatrix-batched-chunks

compilade committed:
imatrix : two-way conversion between old format and GGUF

compilade committed:
convert : remove imatrix to gguf python script

compilade committed:
imatrix : use the function name in more error messages

compilade committed:
Merge branch 'master' into compilade/imatrix-batched-chunks

compilade committed:
imatrix : don't use FMA explicitly

This should make comparisons between the formats easier
because this matches the behavior of the previous version.

compilade committed:
imatrix : avoid returning from void function save_imatrix

compilade committed:
imatrix : support 3d tensors with MUL_MAT

compilade committed:
quantize : fix dataset name loading from gguf imatrix

compilade committed:
Merge branch 'master' into compilade/imatrix-batched-chunks

CISC commented on tools/imatrix/imatrix.cpp:
```diff
@@ -22,31 +24,49 @@
     LOG("\n");
 }
 
+static bool str_has_suffix(const std::string & str, const std::string & suffix) {
+    return str.size() >= suffix.size() && str.compare(str.size() - suffix.size(), str.size(), suffix) == 0;
+}
+
```
```suggestion
```

CISC commented on tools/imatrix/imatrix.cpp:
```diff
@@ -22,31 +24,49 @@
+    return str.size() >= suffix.size() && str.compare(str.size() - suffix.size(), str.size(), suffix) == 0;
+}
+
+static bool str_remove_suffix(std::string & str, const std::string & suffix) {
+    bool has_suffix = str_has_suffix(str, suffix);
+    if (has_suffix) {
+        str = str.substr(0, str.size() - suffix.size());
+    }
+    return has_suffix;
+}
```
```suggestion
static bool str_remove_suffix(std::string & str, const std::string_view & suffix) {
    bool has_suffix = string_ends_with(str, suffix);
    if (has_suffix) {
        str = str.substr(0, str.size() - suffix.size());
    }
    return has_suffix;
}
```
This is a nice complement to `string_ends_with`, should be moved to `common.cpp` as `string_remove_suffix`.

CISC replied:
I did wonder about that, but didn't know enough of the details behind it, reading that it certainly seems so, and it should be worth going over the usage of `std::string_view` elsewhere.

CISC commented on tools/imatrix/imatrix.cpp:
```diff
@@ -259,93 +310,351 @@ void IMatrixCollector::save_imatrix(int ncall) const {
+    auto fname = m_params.out_file;
+
+    // TODO: use the new format by default also for .imatrix
+    if (!str_has_suffix(fname, ".gguf")) {
```
```suggestion
    if (!string_ends_with(fname, ".gguf")) {
```

compilade committed:
common : move string_remove_suffix from quantize and imatrix

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

CISC reviewed (approved):
LGTM, I did some light testing with old and new formats, converting them back and forth, checking that quantization was unchanged, etc, all seems good.

~Only slightly annoying thing is the garbage output to console (by `gguf_init_from_file`) when using old format.~ Resolved in #14381

Will probably benefit from testing by more people before merging, @bartowski1182 @danielhanchen ?

JohannesGaessler commented:
Thank you for working on this, I've been thinking that storing imatrix as gguf wold be nice for investigating the use of gradients instead of activations.

CISC commented:
I guess unless there are any objections from @bartowski1182 or @danielhanchen we can merge this after the weekend?

bartowski1182 commented:
Objections? I've been looking forward to this change for months haha

danielhanchen commented:
Oh this looks fantastic great work! Re using bigger batch sizes - does this mean if memory allows, imatrix should be in fact faster to process via PP?

I'll try this out over the month, but maybe in the meantime, I'll temporarily use the legacy format - but overall this change is very welcome! Great work @compilade !

compilade commented on tools/quantize/quantize.cpp:
```diff
@@ -180,31 +187,148 @@ static int load_imatrix(const std::string & imatrix_file, std::string & imatrix_
+            const float count = ((const float *) counts->data)[j];
+            if (count > 0.0f) {
+                for (int64_t i = 0; i < ne0; ++i) {
+                    e[j*ne0 + i] = ((const float *) sums->data)[j*ne0 + i] / count;
```
The last paragraphs of my previous comment (in <https://github.com/ggml-org/llama.cpp/pull/9400#issuecomment-3043442903>) made me think of <https://github.com/ggml-org/llama.cpp/pull/12557#issuecomment-2751226320> by @jukofyork which suggested the `imatrix` weights calculated from the dataset(s) should gradually replace the prior of "equal weights" (aka using `1` for the `imatrix` weights).

Now I wonder if always using an extra `1` in the sum of squared activations would be sufficient to make the formula closer to act as a regulariser (I might be misusing the term), at least near the lower extreme.

Not sure what relative weight the prior should have, though. Maybe `(1.0f)/(1.0f)` would be too little, since it becomes `1/512` after only one chunk.

(Obviously, this is out of scope for this PR (and will not be included here), but it's still somewhat related, because **the new format stores per-matrice token counts instead of per-tensor chunk counts**, which would make the above possible sanely for stacked MoE tensors (which may not be used equally even within a chunk))

E.g.

```suggestion
                    e[j*ne0 + i] = (((const float *) sums->data)[j*ne0 + i] + 1.0f) / (count + 1.0f);
```

jukofyork replied:
> The last paragraphs of my previous comment (in [#9400 (comment)](https://github.com/ggml-org/llama.cpp/pull/9400#issuecomment-3043442903)) made me think of [#12557 (comment)](https://github.com/ggml-org/llama.cpp/pull/12557#issuecomment-2751226320) by @jukofyork which suggested the `imatrix` weights calculated from the dataset(s) should gradually replace the prior of "equal weights" (aka using `1` for the `imatrix` weights).
> 
> Now I wonder if always using an extra `1` in the sum of squared activations would be sufficient to make the formula closer to act as a regulariser (I might be misusing the term), at least near the lower extreme.

This idea actually goes right back to the roots of Bayesianism:

https://en.wikipedia.org/wiki/Sunrise_problem

or more formally/recently:

https://en.wikipedia.org/wiki/Rule_of_succession

This sort of "pseudocount" was used extensively in the NLP literature in the pre-LLM days:

https://en.wikipedia.org/wiki/Additive_smoothing

(as often you ended up with very small samples in some buckets when performing n-gram analysis, etc)

or more formally, it's the formula for posterior predictive of a Bernoulli likelihood, see the first row of the first table here:

https://en.wikipedia.org/wiki/Conjugate_prior

> Not sure what relative weight the prior should have, though. Maybe `(1.0f)/(1.0f)` would be too little, since it becomes `1/512` after only one chunk.

Since the `imatrix` values are a non-negativite weight centered around 1, a:

https://en.wikipedia.org/wiki/Log-normal_distribution

prior would likely be the best place to start (and is very easy to deal with / program up as it's just the normal posterior predictive formula using log-transformed data - the Wikipedia page above links to a paper that shows this).

danielhanchen replied:
I was experimenting with the new format - great work @compilade ! Re the `(W + 1)/(N + 1)` ie is the motivation specifically for MoE experts where there's little to no activation data? And the goal is to penalize the weights?

compilade replied:
> Re the `(W + 1)/(N + 1)` ie is the motivation specifically for MoE experts where there's little to no activation data? 

@danielhanchen
Yes. It wouldn't cause much difference in other situations, only when the evaluation count (`N`) is small.

> And the goal is to penalize the weights?

Not quite "penalize", more like smoothly transition between the "no calibration" prior, and further calibration.

Otherwise, for the MoE experts which were only used for very few tokens, the activations have a much higher impact on quantization than they should have (since the sample size is small).

I'm not sure what value would be best to use in `(W + v) / (N + v)` (`v` could be `1`, the chunk over the expert count, directly the chunk size, etc.), so I did not implement this here. It would be great to explore if it's beneficial and what value would be ideal.

danielhanchen replied:
Wait you also mentioned `1 / 512` was this implemented previously before this PR? Or is it still doing `1 / 512` if we don't see ample data?

compilade replied:
> Hmm I'm assuming `v = 0.01, 0.1, 1` all make sense - `v > 1` might be good for outliers -> maybe a small `0.01` might be effective as a starter?

@danielhanchen
I think a too-small `v` would not make much difference to the status quo of `v = 0`. **Weights of 1 are neutral (it's what is used when there is no imatrix)**, and so making `(in_sum2 + v) / (counts + v)` tend towards 1 when `counts` is small is the goal.

The evaluation counts are integers (even though stored as F32), and they correspond to the number of tokens processed from the calibration dataset (or at least the number of them which reached the particular tensor).

For example, a `v` of 1 would mean that a `counts` of `1` would affect `1 / 2` of the result, a `counts` of `512` would affect `512 / 513` of the result, etc. It's like a weighted average where `1` has a weight too. Bigger `v` means it takes more tokens to fully consider the imatrix data. (E.g. `v = 256` and `counts = 1024` (2 chunks) would mean the `in_sum2` only affect `1024 / 1280 = 0.8` of the resulting weighted average with `1`)

> Or is it still doing `1 / 512` if we don't see ample data?

The only thing it's doing is using `1` (neutral weights) when `counts` is zero. Not `1 / 512`, `1`. This is like using an infinitesimal (very small) `v`, because `(in_sum2 + v) / (counts + v)` would be 1 when `in_sum2` and `counts` are 0.

saood06 commented:
> Oh, there's another behavior which differs: when a tensor sub-matrice was not solicited by a dataset (e.g. a MoE model with unsolicited experts), the old format skips the entire tensor (even when e.g. 95% of the experts were solicited), while the new format keeps it, and it's in `llama-quantize` that the zeros are handled by using imatrix weights of `1` for that sub-matrice, which from my experiments around #12557, seems reasonable, and mostly matches what the quantization functions do without `imatrix`. This should allow avoiding problems like #12913.
> 
> Partial data is not handled at read time for the legacy format because of insufficient shape information. (It could be [handled at write time](https://github.com/nicoboss/llama.cpp/pull/1) (as in @nicoboss's fork), but that workaround would incorrectly affect merges of MoE `imatrix` files (by adding a `1` value to the squared activations of unused experts, even when those experts are used in another merged `imatrix` file), although arguably the current behavior (dropping the data for the entire tensor) is also wrong. Both problems cannot be fixed simultaneously without a different format than the legacy one.)

The solution in @nicoboss's fork was inspired by https://github.com/ikawrakow/ik_llama.cpp/pull/202 which does mention this concern:

>Strictly speaking it would be better to leave the zeros in the imatrix data of experts that have never been activated. But this would require to go and add proper protection against all-zeros imatrices, along with the appropriate corrective action, for all quants, and not just for IQ1_S_R4 as I did in https://github.com/ikawrakow/ik_llama.cpp/pull/191. So, for now we go with same-importance columns for never activated experts.

compilade committed:
Merge branch 'master' into compilade/imatrix-batched-chunks

compilade committed:
imatrix : add warning when legacy format is written

compilade committed:
imatrix : warn when writing partial data, to help guess dataset coverage

Also make the legacy format store partial data
by using neutral values for missing data.
This matches what is done at read-time for the new format,
and so should get the same quality in case the old format is still used.

compilade committed:
imatrix : avoid loading model to convert or combine imatrix

compilade committed:
imatrix : avoid using imatrix.dat in README

compilade commented:
To address some feedback I got recently, I've added a warning when writing using the legacy format so that it's more obvious what is happening.

```
save_imatrix: saving to legacy imatrix format because output suffix is not .gguf
```

I've also added back the warnings for partial data for the new format, because it can still be useful to know that is happening, even if the data is not omitted (partial data is handled at read-time in `llama-quantize`, this allows both correct `imatrix.gguf` combining *and* weighting missing data neutrally).

And to make the old format a bit more equivalent in quality to the new format (except when combining multiple `imatrix.dat` files with `--in-file`), I've made it write `1` values where the evaluation count is zero, a bit like in [`nicoboss`'s fork](<https://github.com/nicoboss/llama.cpp/pull/1>), but at write time (without affecting the internal data) instead of at collect time. (this is different than my previous stance in the last paragraph of <https://github.com/ggml-org/llama.cpp/pull/9400#issuecomment-3043442903>, because I realized dropping data would also affect combining imatrix files, and since most people don't combine imatrix files, then having the same behavior as the new format in the most common use case is saner)

I've also removed the need to load a model when converting between formats (it was already kind of like this when combining imatrix files), and so the following should be possible:

```console
$ ./bin/llama-imatrix --in-file imatrix.dat -o imatrix.gguf
$ ./bin/llama-imatrix --in-file imatrix.gguf -o imatrix-roundtrip.dat
$ ./bin/llama-imatrix --in-file imatrix-roundtrip.dat -o imatrix-roundtrip.gguf
```

Note that shape information for evaluation counts of MoE tensors is missing from legacy imatrix files, and so it will also be missing from the converted `imatrix.gguf` file, except if more data is provided or if it's merged with a fresh `imatrix.gguf` file of the same model. (it will still work with `llama-quantize`, even when the evaluation count shape is flattened; GGUF makes it easy to support that)

Preserving the shape of evaluation counts is partly why it's recommended to use `.gguf` for newly-generated imatrix files.

The forced suffix of `.gguf` for GGUF-based imatrix files might be controversial, but [a `.gguf` suffix is necessary for HuggingFace to display its GGUF previews](https://github.com/huggingface/huggingface.js/blob/189d0a90dd4acc1e3ea1faa5b73525468c334454/packages/gguf/src/gguf.ts#L18) anyway (even though technically GGUF has a magic header and so it can be identified from its contents). This restriction will likely be removed once writing to the old format isn't supported anymore (in a future PR, not this one).

Since the old format doesn't have a magic header, `llama-quantize` will always try to load imatrix files as GGUF first, and fallback to the legacy format when it fails (this means the filename suffix of imatrix files technically doesn't matter at load time).

nicoboss commented:
@compilade Thanks a lot for your hard work. I'm really looking forward to this PR getting merged! Everything is perfect now in my opinion.

> To address some feedback I got recently, I've added a warning when writing using the legacy format so that it's more obvious what is happening.

Thanks a lot for listening to our feedback and adding this warning. This should be enough to warn users that accidentally still use the legacy `imatrix.dat` file format after this is merged.

> I've also added back the warnings for partial data for the new format, because it can still be useful to know that is happening, even if the data is not omitted (partial data is handled at read-time in llama-quantize, this allows both correct `imatrix.gguf` combining and weighting missing data neutrally).

Thanks a lot! This is super useful to judge the quality of the imatrix and imatrix dataset as a great imatrix dataset should cover more experts than a bad one (unfortunately are always cases where even the best imatrix dataset can't cover them all as training the router to eventually make use of all experts seem quite hard and so was not done well for all MoE models).

> I've made it write 1 values where the evaluation count is zero, a bit like in https://github.com/nicoboss/llama.cpp/pull/1, but without modifying the internal data (and so intermediate saving will not affect the final result).

That's super cool. I hated how my patch broke intermediate saving both by them affecting the result and by them hiding how much experts are covered for future saves and why we had to disable intermediate saves. This is such an elegant solution!

> I've also removed the need to load a model when converting between formats

I really appreciate and find it super cool how easily conversion between the legacy `imatrix.dat` and new `imatrix.gguf` file format is possible. Not only that you went out of your way to make booth backwards and forwards compatibility as good as possible for everyone.

> The forced suffix of .gguf for GGUF-based imatrix files might be controversial, but [a .gguf suffix is necessary for HuggingFace to display its GGUF previews](https://github.com/huggingface/huggingface.js/blob/189d0a90dd4acc1e3ea1faa5b73525468c334454/packages/gguf/src/gguf.ts#L18) anyway (even though technically GGUF has a magic header and so it can be identified from its contents).

With there now being a warning if someone doesn’t specify the `.gguf` suffix I find this design choose acceptable especially given that you have valid backwards compatibility reasons for it to be this  way.

> This restriction will likely be removed once writing to the old format isn't supported anymore (in a future PR, not this one).

I'm looking forward to that. I appreciate that you give everyone time to slowly adopt the new file format. Please just don't forget to eventually drop writing support for the legacy `imatrix.dat` file format and allow arbitrary imatrix file name suffix.

> Since the old format doesn't have a magic header, llama-quantize will always try to load imatrix files as GGUF first, and fallback to the legacy format when it fails (this means the filename suffix of imatrix files technically doesn't matter at load time).

So the `.gguf` suffix is only forced at write time so the file can be rename afterwards as it will always try to load the file as GGUF first. That’s really nice. Not that there really is any reason to not have them end with `.gguf` given that they are GGUF files but great that we can name them any way we want.

EAddario commented:
Echoing @nicoboss' sentiment. This is a very nice enhancement @compilade. Thank you.

I've been testing by running different permutations of options, including roundtrips on each test, and comparing the resulting stats. As far as I can tell, everything checks out!

The only, minor, observation is that when converting an existing file to the new format, a `gguf_init_from_file_impl: invalid magic characters: '????', expected 'GGUF'` warning is displayed. Other than that, it works like a charm!

CISC commented:
> The only, minor, observation is that when converting an existing file to the new format, a `gguf_init_from_file_impl: invalid magic characters: '????', expected 'GGUF'` warning is displayed. Other than that, it works like a charm!

Unavoidable, but a vast improvement from previous behaviour, see #14381. :)

ubergarm commented:
I released a manline imatrix for Kimi-K2-Instruct using this PR here if anyone is looking for it given it is challenging to compute: https://huggingface.co/ubergarm/Kimi-K2-Instruct-GGUF/tree/main/mainline 

Feel free to use it for cooking your own custom mainline quants. I also have a version for ik_llama.cpp if that is your thing.

Might be cool to look inside it with Ed's imatrix stats tool: https://github.com/ggml-org/llama.cpp/pull/12718

Thanks!

EAddario commented:
Assuming no additional changes on this PR, the enhanced version of #12718 is ready to go as soon as this one is merged

compilade commented:
> @compilade Time to merge this (and adapt #12718 afterwards)?

@CISC Sure. I hope I've tested enough edge cases. Will merge at 16:00 UTC on 2025-07-19 (in around 10 hours), to give some buffer for last-minute problems.

---

(sorry for the delayed reply; recently made changes to my home network, now got symmetric fiber Internet)
